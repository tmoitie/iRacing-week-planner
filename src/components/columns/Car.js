// @flow

import * as React from 'react';
import classnames from 'classnames';
import intersection from 'lodash.intersection';
import { useTranslation } from 'react-i18next';
import StarIcon from '../icon/StarIcon';
import CarModal from '../modal/CarModal';

import styles from '../../styles/main.module.scss';
import ClickableCell from './ClickableCell';

type Props = {
  race: {
    carIds: Array<number>,
    carClasses: Array<string>,
    series: string,
  },
  favouriteCars: Array<number>,
  ownedCars: Array<number>,
};

export default function Car({ race, favouriteCars, ownedCars }: Props) {
  const [modalOpen, setModalOpen] = React.useState(false);
  const { t } = useTranslation();

  const openModal = () => {
    setModalOpen(true);
  };

  const knownCar = race.carIds[0] !== null;

  return (
    <>
      <ClickableCell
        className={classnames({
          [styles.success]: knownCar ? intersection(ownedCars, race.carIds).length !== 0 : false,
        })}
        onClick={openModal}
      >
        {knownCar && intersection(favouriteCars, race.carIds).length !== 0 ? (
          <StarIcon />
        ) : null}
        <span> </span>
        {knownCar ? race.carClasses.join(', ') : t('Unknown')}
      </ClickableCell>
      {knownCar ? (
        <CarModal
          isOpen={modalOpen}
          onClose={() => setModalOpen(false)}
          ownedCars={ownedCars}
          favouriteCars={favouriteCars}
          carIds={race.carIds}
          seriesName={race.series}
          seriesId={race.seriesId}
        />
      ) : null}
    </>
  );
}
