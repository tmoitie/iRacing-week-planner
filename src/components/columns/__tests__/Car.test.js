import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, test } from '@jest/globals';
import userEvent from '@testing-library/user-event';
import { TableWrapper } from './ColumnUtils';

import Car from '../Car';

describe('components/columns/Car', () => {
  test('renders correctly', () => {
    render(
      <TableWrapper>
        <Car
          race={{
            carIds: [4, 5],
            carClasses: ['Mazda Cup'],
            series: 'Mazda Cup',
          }}
          favouriteCars={[]}
          ownedCars={[]}
        />
      </TableWrapper>,
    );

    expect(screen.getByTestId('table-row').firstChild).toMatchSnapshot();
  });

  test('renders owned, favourite', () => {
    render(
      <TableWrapper>
        <Car
          race={{
            carIds: [4, 5],
            carClasses: ['Mazda Cup'],
            series: 'Mazda Cup',
          }}
          favouriteCars={[4]}
          ownedCars={[5]}
        />
      </TableWrapper>,
    );

    expect(screen.getByTestId('table-row').firstChild).toMatchSnapshot();
  });

  test('opens modal', async () => {
    const user = userEvent.setup();

    const { baseElement, unmount } = render(
      <TableWrapper>
        <Car
          race={{
            carIds: [4, 5],
            carClasses: ['Mazda Cup'],
            series: 'Mazda Cup',
            seriesId: 105,
          }}
          favouriteCars={[4]}
          ownedCars={[5]}
        />
      </TableWrapper>,
    );

    await user.click(screen.getByText('Mazda Cup'));
    expect(baseElement).toMatchSnapshot();

    await user.click(screen.getByText('Close'));
    expect(baseElement).toMatchSnapshot();

    unmount();
  });
});
