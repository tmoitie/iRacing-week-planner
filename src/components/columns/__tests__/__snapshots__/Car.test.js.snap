// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/columns/Car opens modal 1`] = `
<body>
  <div>
    <table>
      <tbody>
        <tr
          data-testid="table-row"
        >
          <td
            class="success clickableCell"
          >
            <button
              class="cellButton"
              style="min-height: 0;"
              type="button"
            >
              <span
                class="glyphicon glyphicon-star"
                data-testid="StarIcon"
              />
              <span>
                 
              </span>
              Mazda Cup
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    id="modal-cars-105"
  >
    <div
      class="ReactModal__Overlay"
      style="position: fixed; display: block; inset: 0; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        class="Modal__Bootstrap modal-dialog modal-lg"
        data-testid="modal-cars-105"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <button
              aria-label="Close"
              class="close"
              type="button"
            >
              <span
                class="glyphicon glyphicon-remove"
              />
            </button>
            <h4
              class="modal-title"
            >
              Cars for Mazda Cup
            </h4>
          </div>
          <div
            class="modal-body"
            style="max-height: 55vh; overflow-y: auto;"
          >
            <div
              class="container-fluid"
            >
              <div
                class="table-responsive"
              >
                <table
                  class="table"
                  style="font-size: 0.8em;"
                >
                  <thead>
                    <tr>
                      <th>
                        Car
                      </th>
                    </tr>
                  </thead>
                  <tbody />
                </table>
              </div>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              Close
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`components/columns/Car opens modal 2`] = `
<body>
  <div>
    <table>
      <tbody>
        <tr
          data-testid="table-row"
        >
          <td
            class="success clickableCell"
          >
            <button
              class="cellButton"
              style="min-height: 0;"
              type="button"
            >
              <span
                class="glyphicon glyphicon-star"
                data-testid="StarIcon"
              />
              <span>
                 
              </span>
              Mazda Cup
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    id="modal-cars-105"
  />
</body>
`;

exports[`components/columns/Car renders correctly 1`] = `
<td
  class="clickableCell"
>
  <button
    class="cellButton"
    style="min-height: 0;"
    type="button"
  >
    <span>
       
    </span>
    Mazda Cup
  </button>
</td>
`;

exports[`components/columns/Car renders owned, favourite 1`] = `
<td
  class="success clickableCell"
>
  <button
    class="cellButton"
    style="min-height: 0;"
    type="button"
  >
    <span
      class="glyphicon glyphicon-star"
      data-testid="StarIcon"
    />
    <span>
       
    </span>
    Mazda Cup
  </button>
</td>
`;
